CRM/Core/BAO/Setting.php

```diff
@@ -208,6 +211,9 @@ public static function setItems(&$params, $domains = NULL) {
     $fieldsToSet = self::validateSettingsInput($params, $fields);
 
     foreach ($fieldsToSet as $settingField => &$settingValue) {
+      if (self::isEnvironmentSet($settingField)) {
```

Comment:
Same issue here. This check is being done at BAO level for displaying the error in API. If done at self::validateSetting, then it would be applicable to all settings.